include/crow/middlewares/cors.h

```diff
@@ -0,0 +1,159 @@
+        CORSRules& max_age(int max_age)
+        {
+            max_age_ = std::to_string(max_age);
+            return *this;
+        }
+
+        // Enable Access-Control-Allow-Credentials
+        CORSRules& allow_credentials()
+        {
+            allow_credentials_ = true;
+            return *this;
+        }
+
+        // Ignore CORS and don't send any headers
+        void ignore()
+        {
+            ignore_ = true;
+        }
+
+    private:
+        // build comma separated list
+        void add_list_item(std::string& list, const std::string& val)
+        {
+            if (list == "*") list = "";
+            if (list.size() > 0) list += ", ";
+            list += val;
+        }
+
+        // Set header `key` to `value` if it is not set
+        void set_header(const std::string& key, const std::string& value, crow::response& res)
+        {
+            if (value.size() == 0) return;
+            if (!get_header_value(res.headers, key).empty()) return;
+            res.add_header(key, value);
+        }
+
+        // Set response headers
+        void apply(crow::response& res)
+        {
+            if (ignore_) return;
+            set_header("Access-Control-Allow-Origin", origin_, res);
+            set_header("Access-Control-Allow-Methods", methods_, res);
+            set_header("Access-Control-Allow-Headers", headers_, res);
+            set_header("Access-Control-Max-Age", max_age_, res);
+            if (allow_credentials_) set_header("Access-Control-Allow-Credentials", "true", res);
+        }
+
+        bool ignore_ = false;
+        // TODO: support multiple origins that are dynamically selected
+        std::string origin_ = "*";
+        std::string methods_ = "*";
+        std::string headers_ = "*";
+        std::string max_age_;
+        bool allow_credentials_ = false;
+    };
+
+    /// CORSHandler is a global middleware for setting CORS headers.
+    ///
+    /// By default, it sets Access-Control-Allow-Origin/Methods/Headers to "*".
+    /// The default behaviour can be changed with the `global()` cors rule.
+    /// Additional rules for prexies can be added with `prefix()`.
+    struct CORSHandler
+    {
+        struct context
+        {};
+
+        void before_handle(crow::request& /*req*/, crow::response& /*res*/, context& /*ctx*/)
+        {}
+
+        void after_handle(crow::request& req, crow::response& res, context& /*ctx*/)
+        {
+            auto& rule = find_rule(req.url);
+            rule.apply(res);
+        }
+
+        // Handle CORS on specific prefix path
+        CORSRules& prefix(const std::string& prefix)
+        {
+            rules.emplace_back(prefix, CORSRules{});
+            return rules.back().second;
+        }
+
+        // Global CORS policy
+        CORSRules& global()
+        {
+            return default_;
+        }
+
+    private:
+        CORSRules& find_rule(const std::string& path)
+        {
+            for (auto& rule : rules)
```

Comment:
I expect that there won't be more than a couple of prefixes, so linear search is okay. Something like a trie (there is no generic one in CROW) is a definite overkill.